packages/account/src/account.ts

```diff
@@ -1,60 +1,89 @@
 import Web3 from 'web3'
-// import { Accounts } from 'web3-eth-accounts'
+import * as circomlib from 'circomlib'
 import { Account, EncryptedKeystoreV3Json, AddAccount } from 'web3-core'
 import { Field, Point, EdDSA, signEdDSA, verifyEdDSA } from '@zkopru/babyjubjub'
 import { Keystore } from '@zkopru/prisma'
-import { ZkTx, Utxo } from '@zkopru/transaction'
+import { ZkAddress, ZkTx, Utxo } from '@zkopru/transaction'
 import { hexify } from '@zkopru/utils'
+import createKeccak from 'keccak'
 import assert from 'assert'
 
+const poseidonHash = circomlib.poseidon.createHash(6, 8, 57)
+
 export class ZkAccount {
-  private snarkPK: Field
+  private p: Field // spending key
 
-  private ethPK: string
+  private pG: Point // spending key's EdDSA point
 
-  address: string
+  private n: Field // nullifier seed, viewing key
 
-  pubKey: Point
+  private ethPK: string
+
+  ethAddress: string
 
   ethAccount: Account
 
+  zkAddress: ZkAddress // https://github.com/zkopru-network/zkopru/issues/43
+
   constructor(pk: Buffer | string | Account) {
     if (pk instanceof Buffer || typeof pk === 'string') {
       if (pk instanceof Buffer) {
         this.ethPK = hexify(pk, 32)
-        this.snarkPK = Field.fromBuffer(pk)
+        this.p = Field.fromBuffer(pk)
       } else {
         this.ethPK = hexify(pk, 32)
-        this.snarkPK = Field.from(pk)
+        this.p = Field.from(pk)
       }
       const web3 = new Web3()
       this.ethAccount = web3.eth.accounts.privateKeyToAccount(this.ethPK)
     } else {
       this.ethPK = hexify(pk.privateKey, 32)
-      this.snarkPK = Field.from(pk.privateKey)
+      this.p = Field.from(pk.privateKey)
       this.ethAccount = pk
     }
-    this.address = this.ethAccount.address.toLowerCase()
-    this.pubKey = Point.fromPrivKey(this.snarkPK.toHex(32))
+    this.pG = Point.fromPrivKey(this.p.toHex(32))
+    this.ethAddress = this.ethAccount.address.toLowerCase()
+    // https://github.com/zkopru-network/zkopru/issues/34#issuecomment-666988505
+    // Note: viewing key can be derived using another method. This is just for the convenience
+    // to make it easy to restore spending key & viewing key together from a mnemonic source in
+    // a deterministic way
+    this.n = Field.from(
+      createKeccak('keccak256')
+        .update(this.p.toBytes32().toBuffer())
+        .digest(),
+    )
+    const N = Point.fromPrivKey(this.n.toHex(32))
```

Comment:
@barryWhiteHat 
@weijiekoh 
This part is where the wallet creates spending key, viewing key and its address.